rest-server/src/templates/dockerContainerScript.mustache

```diff
@@ -161,6 +161,24 @@ function hdfs_upload_atomically()
   hdfs dfs -mv $tempFolder ${2} || exit 1
 }
 
+#
+# Prepare the output directory on HDFS
+#
+echo "[INFO] TaskRole:${task_role_no} Task:${PAI_TASK_INDEX}"
+if [ ${task_role_no} -eq 0 ] && [ ${PAI_TASK_INDEX} -eq 0 ] ; then
+  # Create the output directory
+  echo "[INFO] Creating output directory: ${PAI_OUTPUT_DIR}"
+  hdfs dfs -mkdir ${PAI_OUTPUT_DIR}
+  echo "[INFO] Output directory ready."
+else
+  # Waiting for the master container to create the output direcotry
+  echo "[INFO] Waiting for output directory: ${PAI_OUTPUT_DIR}"
+  while ! hdfs dfs -test -d ${PAI_OUTPUT_DIR} ; do
+    sleep 10
+  done
+  echo "[INFO] Output directory ready."
```

Comment:
Have you considered let rest server to create the output folder thus we can avoid potentially a lots of hdfs access in each task?

Reply:
Yes I have considered. I prefer this method because of two reasons:

1. The value of `outputDir` URI is specified in `hdfs://<user's hdfs uri>` format, instead of `http://<user's webhdfs uri>` which can be used by WebHDFS, in job config file, thus rest-server can't use this URI to create the folder without falling back to use the `hdfs` command.

2. Container won't introduce too much extra load to HDFS, since all other operations such as ssh preparation and mpi preparation also use the similar method to operate HDFS. In addition, these operations happen within job containers distributed in NMs, so it is not very possible to be a bottleneck of memory utilization.

How do you think?